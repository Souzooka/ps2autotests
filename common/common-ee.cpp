#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
