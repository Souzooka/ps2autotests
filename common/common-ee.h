#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
