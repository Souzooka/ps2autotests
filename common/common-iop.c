#include <stdio.h>
#include <stdarg.h>
#include <xprintf.h>

#define SCHEDF_BUFFER_SIZE 65536

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsnprintf(schedfBuffer + schedfBufferPos, SCHEDF_BUFFER_SIZE, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}
