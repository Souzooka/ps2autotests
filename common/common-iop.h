#ifndef PS2AUTOTESTS_COMMON_IOP_H
#define PS2AUTOTESTS_COMMON_IOP_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _IOP
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

void schedf(const char *format, ...);
void flushschedf();

#endif
