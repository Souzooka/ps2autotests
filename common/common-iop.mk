# _____     ___ ____     ___ ____
#  ____|   |    ____|   |        | |____|
# |     ___|   |____ ___|    ____| |    \    PS2DEV Open Source Project.
#-----------------------------------------------------------------------
# Copyright 2001-2004, ps2dev - http://www.ps2dev.org
# Licenced under Academic Free License version 2.0
# Review ps2sdk README & LICENSE files for further details.
#
# $Id: Makefile.eeglobal_cpp_sample 1525 2009-01-23 00:24:13Z radad $

# Define pefixes for the various toolchain binaries.
include $(COMMON_DIR)/common-defines.mk

ASFLAGS_TARGET = -mcpu=r3000

CFLAGS_TARGET  = -miop
ASFLAGS_TARGET = -march=r3000
LDFLAGS_TARGET = -miop

# include dir
IOP_INCS := -I$(PS2SDK)/iop/include -I$(PS2SDK)/common/include -I$(COMMON_DIR) \
	-I. $(IOP_INCS)
# C compiler flags
IOP_CFLAGS := $(CFLAGS_TARGET) -O2 -G0 $(IOP_INCS) $(IOP_CFLAGS)

# linker flags
IOP_LDFLAGS := $(LDFLAGS_TARGET) -nostdlib -L$(PS2SDK)/iop/lib $(IOP_LDFLAGS)

# asssembler flags
IOP_ASFLAGS := $(ASFLAGS_TARGET) -EL -G0 $(IOP_ASFLAGS)

# link with following libraries (libs need to be defined multiple times in order for linking to work!!)
IOP_LIBS += -lkernel -lgcc

# Externally defined variables: IOP_BIN, IOP_OBJS, IOP_LIB

%.o : %.c
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.S
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.s
	$(IOP_AS) $(IOP_ASFLAGS) $< -o $@

%.o : %.lst
	echo "#include \"irx_imports.h\"" > build-imports.c
	cat $< >> build-imports.c
	$(IOP_CC) $(IOP_CFLAGS) -c build-imports.c -o $@
	rm -f build-imports.c

%.irx : %.o $(EXTRA_OBJS)
	$(IOP_CC) $(IOP_LDFLAGS) -o $@ $< $(EXTRA_OBJS) $(IOP_LIBS)

all: $(TARGETS:=.irx)

clean:
	-$(RM) -f $(TARGETS:=.irx) $(TARGETS:=.o)

rebuild: clean all
