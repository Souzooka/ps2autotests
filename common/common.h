#ifndef PS2AUTOTESTS_COMMON_H
#define PS2AUTOTESTS_COMMON_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#endif
