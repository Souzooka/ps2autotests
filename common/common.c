#include "common.h"
#include <stdarg.h>

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}
