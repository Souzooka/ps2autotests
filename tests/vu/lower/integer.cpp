#include <common.h>
#include "../assemble.h"

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");

	using namespace VU;
	Block block(vu0_micro + 0);
	block.Wr(IADDI(VI01, VI00, 5));
	block.SafeExit();

	register u32 vi01 = 0;
	asm volatile (
		"vcallms 0\n"
		"vnop\n"
		"cfc2 %0, vi01\n"
		: "=r"(vi01)
	);

	printf("Running works: %04x\n", vi01);

	printf("-- TEST END\n");
	return 0;
}
