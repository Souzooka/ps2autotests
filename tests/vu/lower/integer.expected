-- TEST BEGIN
iadd
  iadd 0, 0: 0000 (no flag changes)
  iadd 0, 1: 0001 (no flag changes)
  iadd 1, 1: 0002 (no flag changes)
  iadd 1, 0: 0001 (no flag changes)
  iadd 2, 2: 0004 (no flag changes)
  iadd 65535, 1: 8002 (no flag changes)
  iadd 65535, 65535: 0002 (no flag changes)
  iadd 57005, 65535: 2154 (no flag changes)
  iadd 57005, 3855: b062 (no flag changes)
  iadd CVI_ZERO, CVI_ZERO: 0000 (no flag changes)
  iadd CVI_ZERO, CVI_ONE: 0001 (no flag changes)
  iadd CVI_ONE, CVI_ZERO: 0001 (no flag changes)
  iadd CVI_ONE, CVI_ONE: 0002 (no flag changes)
  iadd CVI_ONE, CVI_NEGONE: 0000 (no flag changes)
  iadd CVI_S16_MAX, CVI_S16_MAX: fffe (no flag changes)
  iadd CVI_S16_MIN, CVI_S16_MIN: 0000 (no flag changes)
  iadd CVI_S32_MAX, CVI_S32_MAX: fffe (no flag changes)
  iadd CVI_S32_MIN, CVI_S32_MIN: 0000 (no flag changes)
  iadd CVI_S64_MAX, CVI_S64_MAX: fffe (no flag changes)
  iadd CVI_S64_MIN, CVI_S64_MIN: 0000 (no flag changes)
  iadd CVI_GARBAGE1, CVI_GARBAGE2: d226 (no flag changes)
  iadd -> $0: 0000

iaddi
  iaddi 0, 0: 0000 (no flag changes)
  iaddi 0, 1: 0001 (no flag changes)
  iaddi 1, 1: 0002 (no flag changes)
  iaddi 0, 15: 000f (no flag changes)
  iaddi 0, -16: fff0 (no flag changes)
  iaddi 1233, 1: 04d2 (no flag changes)
  iaddi CVI_ZERO, 0: 0000 (no flag changes)
  iaddi CVI_ZERO, 1: 0001 (no flag changes)
  iaddi CVI_ONE, 0: 0001 (no flag changes)
  iaddi CVI_ONE, 1: 0002 (no flag changes)
  iaddi CVI_ONE, -1: 0000 (no flag changes)
  iaddi CVI_S16_MAX, 15: 800e (no flag changes)
  iaddi CVI_S16_MIN, -16: 7ff0 (no flag changes)
  iaddi CVI_S32_MAX, 15: 000e (no flag changes)
  iaddi CVI_S32_MIN, -16: fff0 (no flag changes)
  iaddi CVI_S64_MAX, 15: 000e (no flag changes)
  iaddi CVI_S64_MIN, -16: fff0 (no flag changes)
  iaddi CVI_GARBAGE1, 3: 133a (no flag changes)
  iaddi -> $0: 0000

iaddiu
  iaddiu 0, 0: 0000 (no flag changes)
  iaddiu 0, 1: 0001 (no flag changes)
  iaddiu 1, 1: 0002 (no flag changes)
  iaddiu 1, 0: 0001 (no flag changes)
  iaddiu 2, 2: 0004 (no flag changes)
  iaddiu 65535, 1: 8002 (no flag changes)
  iaddiu 65535, 32767: 0000 (no flag changes)
  iaddiu 57005, 32767: 2152 (no flag changes)
  iaddiu 57005, 3855: b062 (no flag changes)
  iaddiu CVI_ZERO, 0: 0000 (no flag changes)
  iaddiu CVI_ZERO, 1: 0001 (no flag changes)
  iaddiu CVI_ONE, 0: 0001 (no flag changes)
  iaddiu CVI_ONE, 1: 0002 (no flag changes)
  iaddiu CVI_ONE, 32767: 8000 (no flag changes)
  iaddiu CVI_S16_MAX, 32767: fffe (no flag changes)
  iaddiu CVI_S16_MIN, 32767: ffff (no flag changes)
  iaddiu CVI_S32_MAX, 32767: 7ffe (no flag changes)
  iaddiu CVI_S32_MIN, 32767: 7fff (no flag changes)
  iaddiu CVI_S64_MAX, 32767: 7ffe (no flag changes)
  iaddiu CVI_S64_MIN, 32767: 7fff (no flag changes)
  iaddiu CVI_GARBAGE1, 4919: 266e (no flag changes)
  iaddiu -> $0: 0000

iand
  iand 0, 0: 0000 (no flag changes)
  iand 0, 1: 0000 (no flag changes)
  iand 1, 1: 0001 (no flag changes)
  iand 1, 0: 0000 (no flag changes)
  iand 2, 2: 0002 (no flag changes)
  iand 65535, 1: 0001 (no flag changes)
  iand 65535, 65535: 8001 (no flag changes)
  iand 57005, 65535: 8001 (no flag changes)
  iand 57005, 3855: 0103 (no flag changes)
  iand CVI_ZERO, CVI_ZERO: 0000 (no flag changes)
  iand CVI_ZERO, CVI_ONE: 0000 (no flag changes)
  iand CVI_ONE, CVI_ZERO: 0000 (no flag changes)
  iand CVI_ONE, CVI_ONE: 0001 (no flag changes)
  iand CVI_ONE, CVI_NEGONE: 0001 (no flag changes)
  iand CVI_S16_MAX, CVI_S16_MAX: 7fff (no flag changes)
  iand CVI_S16_MIN, CVI_S16_MIN: 8000 (no flag changes)
  iand CVI_S32_MAX, CVI_S32_MAX: ffff (no flag changes)
  iand CVI_S32_MIN, CVI_S32_MIN: 0000 (no flag changes)
  iand CVI_S64_MAX, CVI_S64_MAX: ffff (no flag changes)
  iand CVI_S64_MIN, CVI_S64_MIN: 0000 (no flag changes)
  iand CVI_GARBAGE1, CVI_GARBAGE2: 1227 (no flag changes)
  iand -> $0: 0000

ior
  ior 0, 0: 0000 (no flag changes)
  ior 0, 1: 0001 (no flag changes)
  ior 1, 1: 0001 (no flag changes)
  ior 1, 0: 0001 (no flag changes)
  ior 2, 2: 0002 (no flag changes)
  ior 65535, 1: 8001 (no flag changes)
  ior 65535, 65535: 8001 (no flag changes)
  ior 57005, 65535: a153 (no flag changes)
  ior 57005, 3855: af5f (no flag changes)
  ior CVI_ZERO, CVI_ZERO: 0000 (no flag changes)
  ior CVI_ZERO, CVI_ONE: 0001 (no flag changes)
  ior CVI_ONE, CVI_ZERO: 0001 (no flag changes)
  ior CVI_ONE, CVI_ONE: 0001 (no flag changes)
  ior CVI_ONE, CVI_NEGONE: ffff (no flag changes)
  ior CVI_S16_MAX, CVI_S16_MAX: 7fff (no flag changes)
  ior CVI_S16_MIN, CVI_S16_MIN: 8000 (no flag changes)
  ior CVI_S32_MAX, CVI_S32_MAX: ffff (no flag changes)
  ior CVI_S32_MIN, CVI_S32_MIN: 0000 (no flag changes)
  ior CVI_S64_MAX, CVI_S64_MAX: ffff (no flag changes)
  ior CVI_S64_MIN, CVI_S64_MIN: 0000 (no flag changes)
  ior CVI_GARBAGE1, CVI_GARBAGE2: bfff (no flag changes)
  ior -> $0: 0000

isub
  isub 0, 0: 0000 (no flag changes)
  isub 0, 1: ffff (no flag changes)
  isub 1, 1: 0000 (no flag changes)
  isub 1, 0: 0001 (no flag changes)
  isub 2, 2: 0000 (no flag changes)
  isub 65535, 1: 8000 (no flag changes)
  isub 65535, 65535: 0000 (no flag changes)
  isub 57005, 65535: 2152 (no flag changes)
  isub 57005, 3855: 9244 (no flag changes)
  isub CVI_ZERO, CVI_ZERO: 0000 (no flag changes)
  isub CVI_ZERO, CVI_ONE: ffff (no flag changes)
  isub CVI_ONE, CVI_ZERO: 0001 (no flag changes)
  isub CVI_ONE, CVI_ONE: 0000 (no flag changes)
  isub CVI_ONE, CVI_NEGONE: 0002 (no flag changes)
  isub CVI_S16_MAX, CVI_S16_MAX: 0000 (no flag changes)
  isub CVI_S16_MIN, CVI_S16_MIN: 0000 (no flag changes)
  isub CVI_S32_MAX, CVI_S32_MAX: 0000 (no flag changes)
  isub CVI_S32_MIN, CVI_S32_MIN: 0000 (no flag changes)
  isub CVI_S64_MAX, CVI_S64_MAX: 0000 (no flag changes)
  isub CVI_S64_MIN, CVI_S64_MIN: 0000 (no flag changes)
  isub CVI_GARBAGE1, CVI_GARBAGE2: 5448 (no flag changes)
  isub -> $0: 0000

isubiu
  isubiu 0, 0: 0000 (no flag changes)
  isubiu 0, 1: ffff (no flag changes)
  isubiu 1, 1: 0000 (no flag changes)
  isubiu 1, 0: 0001 (no flag changes)
  isubiu 2, 2: 0000 (no flag changes)
  isubiu 65535, 1: 8000 (no flag changes)
  isubiu 65535, 32767: 0002 (no flag changes)
  isubiu 57005, 32767: 2154 (no flag changes)
  isubiu 57005, 3855: 9244 (no flag changes)
  isubiu CVI_ZERO, 0: 0000 (no flag changes)
  isubiu CVI_ZERO, 1: ffff (no flag changes)
  isubiu CVI_ONE, 0: 0001 (no flag changes)
  isubiu CVI_ONE, 1: 0000 (no flag changes)
  isubiu CVI_ONE, 32767: 8002 (no flag changes)
  isubiu CVI_S16_MAX, 32767: 0000 (no flag changes)
  isubiu CVI_S16_MIN, 32767: 0001 (no flag changes)
  isubiu CVI_S32_MAX, 32767: 8000 (no flag changes)
  isubiu CVI_S32_MIN, 32767: 8001 (no flag changes)
  isubiu CVI_S64_MAX, 32767: 8000 (no flag changes)
  isubiu CVI_S64_MIN, 32767: 8001 (no flag changes)
  isubiu CVI_GARBAGE1, 4919: 0000 (no flag changes)
  isubiu -> $0: 0000

-- TEST END
